dbt/adapters/base/connections.py

```diff
@@ -0,0 +1,261 @@
+import abc
+import multiprocessing
+
+import six
+
+import dbt.exceptions
+import dbt.flags
+from dbt.compat import abstractclassmethod
+from dbt.contracts.connection import Connection
+from dbt.logger import GLOBAL_LOGGER as logger
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseConnectionManager(object):
```

Comment:
how is thread safety / lack of thread safety typically annotated in Python programs? there are a number of threadsafe methods here, and a handful that aren't. is there a good way to distinguish them?

Reply:
Comments, sadly. It's probably better to make these things thread-safe. Is `get()` the main issue? This is really another point in favor of tying connections directly to threads.

Reply:
> make these things thread-safe

using a re-entrant lock?

> Is get() the main issue?

I have issues with how some of these things are implemented, but this looks like a faithful refactor of what existed before in the old inheritance structure. I do not want to block this PR on that

> This is really another point in favor of tying connections directly to threads.

Thread-local connections? Or a separate threadpool managed by the connection manager? I'd be interested to hear more about how this might work. Note that it's very important not to open more than `num_threads + buffer` connections to end user warehouses.